### PR TITLE
Issue 46 Define model for notification preferences

### PR DESCRIPTION
**Pull request by heeylena**
In this task, I defined appropriate models to allow users to customize their notification preferences.

#### Models

- **NotificationType** — Stores different types of notifications.
- **UserNotificationPreference** — Global notification settings per user.
- **UserNotificationTypePreference** — Per-notification-type preferences.

##### Get/Update User Notification Preferences
- **GET** `/api/v1/communications/preferences/` — Get the current user's notification preferences.  
- **PATCH** `/api/v1/communications/preferences/` — Update notification preferences.

##### Get Available Notification Types
- **GET** `/api/v1/communications/notification-types/` — List all available notification types.

##### Get Notification Preference Options
- **GET** `/api/v1/communications/preferences/options/` — Get available options, like frequency choices.

##### Update Specific Notification Type Preference
- **PATCH** `/api/v1/communications/preferences/{id}/update_type_preference/` — Update a specific notification type's preference.

All endpoints require authentication and will only return/update the preferences of the currently logged-in user.
Additionally, I added the ability to manage notification types and user preferences through the Django admin interface.
All tests passed successfully:
Image 14 08 2025 at 8 33 PM

### CODE REVIEW

File: communications/admin.py [hunk 1]
```diff
@@ -1,3 +1,59 @@
...
+class NotificationTypeAdmin(admin.ModelAdmin):
+    """Admin interface for NotificationType model."""
+    list_display = ('name', 'code', 'is_active', 'created_at')
+    list_filter = ('is_active',)
+    search_fields = ('name', 'code', 'description')
+    readonly_fields = ('created_at', 'updated_at')
+    fieldsets = (
+        (None, {
+            'fields': ('code', 'name', 'description', 'is_active')
+        }),
```
Comment (mehalyna) [inline communications/admin.py:19]: For better maintainability, consider using list comprehensions or `model._meta.get_fields()` if you have many fields, instead of hardcoding field names.

File: communications/admin.py [hunk 2]
```diff
@@ -1,3 +1,59 @@
...
+    readonly_fields = ('created_at', 'updated_at')
+    inlines = [UserNotificationTypePreferenceInline]
+    
+    fieldsets = (
+        (_('User'), {
+            'fields': ('user',)
+        }),
+        (_('Notification Channels'), {
+            'fields': ('enable_in_app', 'enable_email', 'enable_push')
+        }),
```
Comment (mehalyna) [inline communications/admin.py:49]: Grouping notification channel fields is clear, but if you add more channels in the future, consider using a list for easier extension.

File: communications/models.py [hunk 3]
```diff
@@ -1,3 +1,192 @@
...
+from django.conf import settings
+from django.utils.translation import gettext_lazy as _
 
-# Create your models here.
+
+class NotificationFrequency(models.TextChoices):
+    IMMEDIATE = 'immediate', _('Immediate')
+    DAILY_DIGEST = 'daily_digest', _('Daily Digest')
+    WEEKLY_SUMMARY = 'weekly_summary', _('Weekly Summary')
+    DISABLED = 'disabled', _('Disabled')
```
Comment (mehalyna) [inline communications/models.py:10]: The `NotificationFrequency` choices are duplicated later in `NotificationType.default_frequency` and `UserNotificationTypePreference.frequency`. To avoid inconsistencies, use `NotificationFrequency.choices` everywhere instead of redefining the choices list.

File: communications/models.py [hunk 4]
```diff
@@ -1,3 +1,192 @@
...
+        max_length=20,
+        choices=[
+            ('immediate', _('Immediate')),
+            ('daily_digest', _('Daily Digest')),
+            ('weekly_summary', _('Weekly Summary')),
+            ('disabled', _('Disabled'))
+        ],
+        default='immediate',
+        help_text=_('Default frequency for this notification type')
+    )
```
Comment (mehalyna) [inline communications/models.py:38]: For default_frequency, use `choices=NotificationFrequency.choices` rather than redefining the list of choices for maintainability.

File: communications/models.py [hunk 5]
```diff
@@ -1,3 +1,192 @@
...
+                'name': 'Default Notification',
+                'description': 'Default notification type',
+            }
+        )
+        return obj.pk
+
+
+class Notification(models.Model):
+    """Model to store all notifications sent to users."""
+    notification_id = models.UUIDField(unique=True, editable=False, primary_key=True)
```
Comment (mehalyna) [inline communications/models.py:75]: Consider using `UUIDField` with `default=uuid.uuid4` to automatically generate UUIDs if your logic allows. Otherwise, ensure that `notification_id` is set at creation.

File: communications/models.py [hunk 6]
```diff
@@ -1,3 +1,192 @@
...
+    triggered_by_type = models.CharField(
+        max_length=20,
+        choices=[
+            ('investor', _('Investor')),
+            ('startup', _('Startup')),
+            ('system', _('System'))
+        ],
+        null=True,
+        blank=True
+    )
```
Comment (mehalyna) [inline communications/models.py:105]: Using `triggered_by_user` and `triggered_by_type` together can be confusing; clarify in comments or documentation when each should be used. Consider an enum or a more explicit relationship for `triggered_by_type`.

File: communications/models.py [hunk 7]
```diff
@@ -1,3 +1,192 @@
...
+    class Meta:
+        ordering = ['-created_at']
+        indexes = [
+            models.Index(fields=['user', 'is_read', 'created_at'], name='idx_notif_user_read'),
+            models.Index(fields=['user', 'notification_type', 'created_at'], name='idx_notif_user_type'),
+            models.Index(fields=['triggered_by_user'], name='idx_notif_triggered_by'),
+            models.Index(fields=['related_startup_id'], name='idx_notif_startup'),
+            models.Index(fields=['related_project_id'], name='idx_notif_project'),
+            models.Index(fields=['expires_at'], name='idx_notif_expires'),
+        ]
```
Comment (mehalyna) [inline communications/models.py:134]: You have a lot of indexes; review your most common queries to avoid over-indexing and slowing down writes.

File: communications/models.py [hunk 8]
```diff
@@ -1,3 +1,192 @@
...
+    frequency = models.CharField(
+        max_length=20,
+        choices=NotificationFrequency.choices,
+        default=NotificationFrequency.IMMEDIATE
+    )
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        unique_together = ('user_preference', 'notification_type')
```
Comment (mehalyna) [inline communications/models.py:188]: `unique_together` is now deprecated in favor of `UniqueConstraint` in `Meta.constraints`. Update to use `UniqueConstraint` for forward compatibility.

File: communications/serializers.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    UserNotificationTypePreference
+)
+
+
+class NotificationTypeSerializer(serializers.ModelSerializer):
+    """Serializer for notification types."""
+    class Meta:
+        model = NotificationType
+        fields = ['id', 'code', 'name', 'description', 'is_active']
+        read_only_fields = fields
```
Comment (mehalyna) [inline communications/serializers.py:15]: Setting read_only_fields = fields makes all fields read-only, which may be too restrictive if you ever want to use this serializer for creating or updating NotificationType. Consider explicitly listing only the fields that should be read-only.

File: communications/serializers.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            'notification_id',
+            'notification_type',
+            'title',
+            'message',
+            'is_read',
+            'priority',
+            'priority_display',
+            'created_at',
+            'expires_at',
+            'action_link',
```
Comment (mehalyna) [inline communications/serializers.py:111]: action_link is included in fields but is not defined in the serializer or the Notification model (at least in the provided code). This will raise an error unless action_link is a property or method on the model.

File: communications/serializers.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        ]
+        read_only_fields = [
+            'notification_id', 'created_at', 'updated_at', 'is_read'
```
Comment (mehalyna) [inline communications/serializers.py:114]: is_read is included in read_only_fields, but if you want to allow marking a notification as read via the API, remove it from read_only_fields or provide a separate endpoint/serializer for marking as read.

File: communications/serializers.py
Comment (mehalyna) [inline communications/serializers.py:1]: Add docstrings to all serializer classes for clarity and maintainability, especially if you plan to extend them.

File: communications/signals.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+from .models import UserNotificationPreference, NotificationType, UserNotificationTypePreference
+
+
+@receiver(post_save, sender=settings.AUTH_USER_MODEL)
+def create_user_notification_preferences(sender, instance, created, **kwargs):
+    """
+    Create default notification preferences and type preferences when a new user is created.
+    """
+    if created:
+        preference = UserNotificationPreference.objects.create(user=instance)
```
Comment (mehalyna) [inline communications/signals.py:13]: UserNotificationPreference.objects.create(user=instance) will raise an IntegrityError if a preference already exists (e.g., if user creation is retried). Consider using get_or_create to make this idempotent.

File: communications/signals.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        
+        notification_types = NotificationType.objects.filter(is_active=True)
+        for notification_type in notification_types:
+            UserNotificationTypePreference.objects.create(
+                user_preference=preference,
+                notification_type=notification_type,
+                frequency=notification_type.default_frequency
+            )
```
Comment (mehalyna) [inline communications/signals.py:21]: When creating UserNotificationTypePreference for each type, check that duplicates are not accidentally created on repeated user creation (see above). Use get_or_create here as well for safety.

File: communications/signals.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@receiver(post_migrate)
```
Comment (mehalyna) [inline communications/signals.py:24]: The @receiver(post_migrate) signal will run on every migration. This can be noisy if not scoped. The check sender.name == 'communications' helps, but you may want to further guard against repeated creation of the same NotificationType records.

File: communications/signals.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                'description': 'Notification when a user receives a new message',
+                'default_frequency': 'immediate',
+            },
+            {
+                'code': 'project_updated',
+                'name': 'Project Updated',
+                'description': 'Notification when a followed project is updated',
+                'default_frequency': 'daily_digest',
+            },
+        ]
```
Comment (mehalyna) [inline communications/signals.py:58]: The notification_types list is hardcoded. If notification types are expected to change or be configurable, consider loading from a fixture or settings.

File: communications/views.py [hunk 16]
```diff
@@ -1,12 +1,148 @@
...
+        """Mark a notification as read."""
+        notification = self.get_object()
+        notification.is_read = True
+        notification.save()
+        return Response({'status': 'notification marked as read'})
+    
+    @action(detail=False, methods=['post'])
+    def mark_all_as_read(self, request):
+        """Mark all notifications as read for the current user."""
+        updated = self.get_queryset().filter(is_read=False).update(is_read=True)
```
Comment (mehalyna) [inline communications/views.py:54]: The update for mark_all_as_read does not update the updated_at timestamp unless your model does so automatically. If you want to track when notifications were read, consider updating the timestamp or adding a field for read_at.

File: communications/views.py [hunk 17]
```diff
@@ -1,12 +1,148 @@
...
+    Requires authentication.
+    """
+    queryset = NotificationType.objects.filter(is_active=True)
+    serializer_class = NotificationTypeSerializer
+    permission_classes = [permissions.IsAuthenticated]
+    pagination_class = None
+    
+    def get_queryset(self):
+        """Return only active notification types."""
+        return NotificationType.objects.filter(is_active=True)
```
Comment (mehalyna) [inline communications/views.py:70]: The get_queryset override in NotificationTypeViewSet duplicates the queryset attribute. You can remove either the attribute or the method for simplicity.

File: communications/views.py [hunk 18]
```diff
@@ -1,12 +1,148 @@
...
+        return obj
+
+    @action(detail=True, methods=['patch'])
+    def update_type_preference(self, request, pk=None):
+        """
+        Update a specific notification type preference.
+        Expected payload: {"notification_type_id": 1, "frequency": "immediate"}
+        """
+        preference = self.get_object()
+        notification_type_id = request.data.get('notification_type_id')
```
Comment (mehalyna) [inline communications/views.py:113]: notification_type_id is taken directly from the request, but not validated as an integer or checked for existence. Add input validation and/or serializer validation for better error handling.

File: communications/views.py [hunk 19]
```diff
@@ -1,12 +1,148 @@
...
+        frequency = request.data.get('frequency')
```
Comment (mehalyna) [inline communications/views.py:114]: frequency is not validated against allowed choices. Before saving, check that the value is in NotificationFrequency.choices. Otherwise, it could result in an invalid state or error.

File: communications/views.py [hunk 20]
```diff
@@ -1,12 +1,148 @@
...
+        if not notification_type_id or not frequency:
+            return Response(
+                {'error': 'notification_type_id and frequency are required'},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        
+        try:
+            type_pref = preference.type_preferences.get(
+                notification_type_id=notification_type_id
+            )
```
Comment (mehalyna) [inline communications/views.py:125]: preference.type_preferences.get(notification_type_id=notification_type_id) will throw an error if notification_type_id is not a valid integer or does not exist. Add try/except for ValueError or TypeError, or perform validation first.

File: communications/views.py [hunk 21]
```diff
@@ -1,12 +1,148 @@
...
+        except UserNotificationTypePreference.DoesNotExist:
+            return Response(
+                {'error': 'Notification type preference not found'},
+                status=status.HTTP_404_NOT_FOUND
+            )
+
+    @action(detail=False, methods=['get'])
+    def options(self, request):
+        """Return available options for notification preferences."""
+        from .models import NotificationFrequency
```
Comment (mehalyna) [inline communications/views.py:141]: The options action overlaps with the default OPTIONS method provided by DRF. Consider renaming this action to avoid confusion with HTTP OPTIONS, or ensure that its purpose is well documented.

File: communications/views.py
Comment (mehalyna) [inline communications/views.py:1]: - All viewsets lack explicit throttling. If exposed publicly, consider adding throttling or rate limiting.
- Add logging for actions that mutate data (e.g., mark_as_read, mark_all_as_read, update_type_preference) for auditability and debugging.
- Consider using DRF's built-in permission classes or custom permissions to further restrict access where needed. Right now, only IsAuthenticated is enforced.
- Add docstrings and comments to all custom actions and methods, especially where expected payloads or responses deviate from DRF conventions.

File: tests/communications/test_notification_preferences.py [hunk 22]
```diff
@@ -0,0 +1,156 @@
+from django.urls import reverse
+from rest_framework import status
+from rest_framework.test import APITestCase
+from django.contrib.auth import get_user_model
+from rest_framework.authtoken.models import Token
+from communications.models import NotificationType, UserNotificationPreference, UserNotificationTypePreference
+
+import logging
+logging.disable(logging.CRITICAL)
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:9]: Disabling logging globally with logging.disable(logging.CRITICAL) will suppress all critical logs during the test run, which may hide important errors or warnings. Consider using Django's override_settings or patching logging only for specific noisy loggers.

File: tests/communications/test_notification_preferences.py [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+            name='Test Type 1',
+            description='Test Description 1',
+            default_frequency='immediate'
+        )
+        self.notification_type2 = NotificationType.objects.create(
+            code='test_type_2',
+            name='Test Type 2',
+            description='Test Description 2',
+            default_frequency='daily_digest'
+        )
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:36]: You manually create NotificationType objects. If your NotificationType creation is handled elsewhere (signals, migrations), this could cause duplicate entries or IntegrityErrors due to unique constraints on code. Consider using get_or_create or clearing out NotificationType before each test.

File: tests/communications/test_notification_preferences.py [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+    def test_get_notification_types(self):
+        """Test retrieving notification types."""
+        url = reverse('communications:notification-type-list')
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:40]: reverse('communications:notification-type-list') assumes that the app_name is set as "communications" and the router is properly registered. If the router or app_name changes, this test will fail. Consider using DRF's router.urls for more robust resolution.

File: tests/communications/test_notification_preferences.py [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+    def test_get_user_preferences(self):
+        """Test retrieving user notification preferences."""
+        url = reverse('communications:user-notification-preference-list')
+        response = self.client.get(url)
+        
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIsInstance(response.data, list)
+        self.assertTrue(len(response.data) > 0)
+        
+        user_prefs = response.data[0]
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:58]: user_prefs = response.data[0] assumes the first item is always the correct user. If the queryset is not filtered to the current user in the view, this could test the wrong object.

File: tests/communications/test_notification_preferences.py [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+        
+        pref.refresh_from_db()
+        self.assertTrue(pref.enable_in_app)
+        self.assertFalse(pref.enable_email)
+        self.assertTrue(pref.enable_push)
+
+    def test_update_notification_type_preference(self):
+        """Test updating a specific notification type preference."""
+        pref = UserNotificationPreference.objects.get(user=self.user)
+        type_pref = pref.type_preferences.first()
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:94]: pref.type_preferences.first() assumes at least one type preference exists. If setup or signal creation fails, this will raise an AttributeError. Add assertions to confirm type_preferences.count() > 0 before accessing first().

File: tests/communications/test_notification_preferences.py [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+        
+        url = reverse(
+            'communications:user-notification-preference-update-type-preference',
+            kwargs={'pk': self.user.pk}
+        )
+        
+        data = {
+            'notification_type_id': type_pref.notification_type.id,
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:102]: 'notification_type_id': type_pref.notification_type.id assumes notification_type exists and is valid. If type_pref is None, this will raise an error.

File: tests/communications/test_notification_preferences.py [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+            password='testpass123',
+            first_name='New',
+            last_name='User'
+        )
+        
+        self.assertTrue(hasattr(new_user, 'notification_preferences'))
+        self.assertEqual(
+            new_user.notification_preferences.type_preferences.count(),
+            NotificationType.objects.count()
+        )
```
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:156]: assertEqual(new_user.notification_preferences.type_preferences.count(), NotificationType.objects.count()) assumes all notification types are active and counted. If NotificationType filtering changes (e.g., is_active=False for some types), this check may not match intended business logic.

File: tests/communications/test_notification_preferences.py
Comment (mehalyna) [inline tests/communications/test_notification_preferences.py:1]: - Add docstrings to all test methods for clarity, especially for those testing edge or error conditions.
- Use factories (e.g., with factory_boy) for model creation to simplify test setup and increase maintainability.
- Consider parameterizing tests where possible to reduce code repetition for similar scenarios.
- Clean up any test data (especially if you use non-transactional test runners or external services).

**Review by mehalyna — CHANGES_REQUESTED** [review]

**Review by KarakaticaAsh — APPROVED** [review]
I like your implementation of notifications, especially nice to see API endpoints listed.
Please, add doc strings to all tests methods for better further maintainability.

File: communications/admin.py (hunk 1 above)
Comment (heeylena) [reply communications/admin.py]: Replaced the hardcoded fieldsets with a get_fieldsets method.

File: communications/admin.py (hunk 2 above)
Comment (heeylena) [reply communications/admin.py]: Created a NOTIFICATION_CHANNELS list constant.

File: communications/models.py (hunk 3 above)
Comment (heeylena) [reply communications/models.py:21]: Replaced the hardcoded choices list with choices=NotificationFrequency.choices

File: communications/models.py (hunk 4 above)
Comment (heeylena) [reply communications/models.py]: Replaced the hardcoded choices list with choices=NotificationFrequency.choices

File: communications/models.py (hunk 6 above)
Comment (heeylena) [reply communications/models.py]: Added usage guidelines documentation as model docstring.

File: communications/serializers.py (hunk 9 above)
Comment (heeylena) [reply communications/serializers.py]: 've updated the NotificationTypeSerializer to make only the id and code fields read-only.

File: communications/signals.py (hunk 12 above)
Comment (heeylena) [reply communications/signals.py]: I updated communications/signals.py to avoid IntegrityError on retries and make the signal idempotent.

File: communications/signals.py (hunk 14 above)
Comment (heeylena) [reply communications/signals.py:30]: Added module-level _types_seeded to prevent repeated seeding during a single migrate run.

File: communications/views.py (hunk 18 above)
Comment (heeylena) [reply communications/views.py:131]: Added input validation and error handling

**Review by Raidriar28 — APPROVED** [review]
Great job implementing a clean and extensible notification preference system!
Suggestion: consider adding validation logic to the Notification model for triggered_by_user vs triggered_by_type, as described in the docstring — enforcing it at the model level could prevent misuse.
